### PR TITLE
[Bug Fix] None and Null Partititon Key Improvements

### PR DESCRIPTION
**Pull request by tvaron3**
## Context
- The `PartitionKey` class is actually for the partition key definition while creating a container and NOT for the partition key values. Partition key values need to be set in the `body` or `partititon_key` of the relevant api.
### Three Special Partition key Values 
1. Null Partition Key
    - In python, this can be achieved by performing a write with `None` as the partition key. The sdk will convert the `None` into `null`. In this pr also added the `NullParttionKeyValue` sentinel as an alternative to achieve this.
 ```python 
{
id: some_id, 
pk: null
}
```
2. Empty Partition Key
    - This can be created using the `NoneParttionKeyValue` sentinel, and then based on the `system` property returned by the gateway for the container properties the sdk will create an empty partition key or a undefined partition key. 
 ```python 
{
id: some_id, 
pk: []
}
```
3. Undefined Partition Key
    - This can be created using the `NoneParttionKeyValue` sentinel, and then based on the `system` property returned by the gateway for the container properties the sdk will create an empty partition key or a undefined partition key. 
 ```python 
{
id: some_id, 
pk: [{}]
}
```

## Goals
1.  Improve consistency across apis for null partition keys and absent partition keys 
    - For null partition keys, I added a new partition key value called `NullPartitionKeyValue` which can be used in the container apis including`query_items` and `query_conflicts`. Passing in `None` works for all the container apis except`query_items` and `query_conflicts` because don't want to break customers that rely on this scenario being a cross partition query. 
    - For absent partition keys in a document, `NonePartitionKeyValue` should be used. This works for all container apis except `change feed`. Here is the issue to follow up on change feed behavior https://github.com/Azure/azure-sdk-for-python/issues/42799
5.  Add test coverage for null partition keys and absent partition keys across all container apis
     - Added test coverage for all container apis across sync and async
6.  Improve api documentation around using `None` as the partition key value
     - Added a file in our docs giving example of the different parition key values and explaining when each should be used. Updated api documentation for operations that accept a `partition_key`
7. Address duplicate argument exception when passing none partition key value to `query_items`
     - Fixed the bug where partition key is `None` would break query items because of a duplicate argument error by always popping the kwargs `partition_key`. We broke the behavior in 4.14.0b2 and previously to that version the behavior was that if `None` partition key value was passed into the `query_items` api then a cross partition query would be performed. I decided to keep that behavior to prevent breaking customers.

### CODE REVIEW

**Review by kushagraThapar — APPROVED** [review]

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py [hunk 1]
```diff
@@ -76,15 +76,15 @@
...
+    _Empty,
+    _build_partition_key_from_properties
 )
 from ._auth_policy_async import AsyncCosmosBearerTokenCredentialPolicy
 from .._cosmos_http_logging_policy import CosmosHttpLoggingPolicy
 from .._range_partition_resolver import RangePartitionResolver
 
 
-PartitionKeyType = Union[str, int, float, bool, Sequence[Union[str, int, float, bool, None]], Type[NonePartitionKeyValue]]  # pylint: disable=line-too-long
+PartitionKeyType = Union[str, int, float, bool, Sequence[Union[str, int, float, bool, None]], _Empty, _Undefined, None]  # pylint: disable=line-too-long
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py]: is there a reason we don't include the `NullPartitionKeyValue` here like we do in other places? what about the sync client_connection?

**Review by simorenoh — COMMENTED** [review]
Just one question, other than that LGTM! thanks Tomas!

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py (hunk 1 above)
Comment (tvaron3) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py]: Yeah I really just changed it to make mypy happy. It doesn't apply to sync because in sync we use _PartitionKeyType from partition_key.py and this includes all the types like `NullPartitionKeyValue` `NonePartitionKeyValue`. I changed async to follow this pattern as well.
